### PR TITLE
[docs] Add ESLint Rule to Prevent Usage of localStorage and sessionStorage

### PR DESCRIPTION
**Pull request by danstepanov**
# Why

Sentry is getting spammed with the [an error](https://sentry.io/organizations/expoio/issues/2555130033/events/2e1e7325f7e94ef6863d85281a716613/?frontend_events=%7B%22event_name%22%3A%22Sign+Up%22%2C%22event_label%22%3A%22google%22%7D&project=1526800) when attempting to access localStorage.getItem(). This leads to a 404 page which is a poor user experience.

Here's how reporting errors to Sentry works, that is how we use `localStorage` when talking to Sentry (`sentry-utilities.ts`)
Screen Shot 2022-04-27 at 11 42 38 PM

# How

Add ESLint rule that prevents the usage of localStorage and sessionStorage. 

A few questions/concerns:
- With this approach, do we want to completely get rid of our use of `localStorage` in `./docs/common/sentry-utilities.ts`?
  -  Answer: No, just eslint ignore the file and prevent future use of `localStorage`
- Given the above error from Sentry, it seems like we should just check whether `localStorage.getItem` exists before calling it, rather than removing `localStorage`, altogether.
  - Answer: Ideally we would do without `localStorage` as it is a fairly hacky solution and was implemented as a hotfix to spam problem in the past. However, based on an initial review of the docs, Sentry does not currently offer this desired type of filtration of events. If we do use `localStorage.getItem`, I do still think we should check for it in the places that it is invoked.
- If we'd like to commit to disabling `localStorage` and `sessionsStorage` then should the relevant code in `sentry-utilities` be removed or ignored? Ignoring the existing code will not resolve the current issue so we'd have to completely remove all of how we currently filter Sentry methods. A further alternative would be to use a persistent state that does everything that we use `localStorage` for atm.
  - Answer: Ignore it. 
- Lastly, why does `isLocalStorageAvailable` contain a try/catch block? There is nothing async or promise-based in the method definition. Perhaps I'm missing something?
  - Answer: localStorage is pretty wonky to detect across browsers. 

# Test Plan

1. Create a new Sentry account and replace the dsn, try spamming a thrown error.
2. Try putting in a use of localStorage anywhere in the code and run `yarn lint`
3. ~As a note, `yarn lint` will currently fail as all instance of localStorage have not yet been removed from `sentry-utilities.ts`. This will be resolved once the desired approach is selected.~

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by byCedric** [comment]
If we can do without it, I think we are better off.

> Ignoring the existing code will not resolve the current issue so we'd have to completely remove all of how we currently filter Sentry methods

Exactly, we only prevent new functionality related to this. But it doesn't help solve this issue.

Might be worth checking if the newer Sentry libraries already does this for us. The docs had Sentry implemented a while ago, so I guess that's where the complex integration comes from.

**Comment by danstepanov** [comment]
After some cursory search regarding Sentry's current functionality, it seems like we can accomplish our intentions with filtering Sentry errors via built-in functionality from Sentry. Particularly, [issue grouping](https://docs.sentry.io/product/data-management-settings/event-grouping/) and [inbound filters for error messages](https://docs.sentry.io/product/data-management-settings/filtering/#error-message). This covers two of the three things that `preprocessSentryError()` currently does. I'm currently testing whether we even need the bit that checks localstorage for existing messages.

**Comment by danstepanov** [comment]
@byCedric 
Preface: When deciding how to implement Sentry, I'm assuming that the driving KPI is the Sentry quota, that is, the monthly number of Events that we pay Sentry to track. This would be why we would be concerned with a spike in reported Events. Subsequently, ensuring Events don't count toward our quota is the same as filtering out reported Events. If this is not the case, please let me know.

TLDR: I'm going to keep the rate limiting code, most of the "known errors to ignore" code (but will enable some stuff in project settings that does the same thing), keep localStorage code but have eslint ignore it.

## Rate Limiting
Reporting of an Event and the notification of an Issue are separate concepts in Sentry. Events are singular instances of an Issue and have a "fingerprint". Issues are groupings of Events (that all share a fingerprint). What we see in the dashboard are Issues. If we'd like to explicitly keep each Issue as only having one Event in the Sentry dashboard, we should keep [the existing code](https://github.com/expo/expo/blob/49009baaf0864567454db2fadd26fcef02e98811/docs/common/sentry-utilities.ts#L86). This prevents regressions of Events being reported, although Sentry does group them into Issues, by default.

By default, Sentry will only notify you of an Issue [once every 30 minutes](https://sentry.io/resources/alert-rules/#exploring-the-defaults) that an Event comes in. You'll still see a count of Events in the dashboard but you won't be notified. So keeping the existing code will literally only add the bit about fewer occurrences of the same (already grouped) Issue in the dashboard. Rate limiting is default already implemented, as far as notifications are concerned.

## Known Errors to Discard
Error message filtering allows us to fully rid ourselves of [this part](https://github.com/expo/expo/blob/49009baaf0864567454db2fadd26fcef02e98811/docs/common/sentry-utilities.ts#L41) of the sentry error preprocessor. There are some options [here](https://sentry.io/settings/test-vpx/projects/nextjs/filters/data-filters/) to filter inbound messages which covers some of our use, however to cover all of our error message filtering, we would need to upgrade to the Business Plan, [we're currently on the Team Plan.](https://sentry.io/pricing/). Upgrading for the purpose of this coverage alone is not worth it, in my opinion, we can just continue to filter using the existing array and turn on some browser extension and legacy browser filters in our current plan.
Screen Shot 2022-04-28 at 1 19 25 PM

## Checking localStorage for the same Error within the last 24 hours
I have searched through the Sentry docs and have yet to find something that performs exactly what this chunk handles. I suggest we keep it but set eslint to ignore it.

**Comment by danstepanov** [comment]
Please note: once this PR is approved to be merged, we will need to [configure Sentry](https://sentry.io/settings/test-vpx/projects/nextjs/filters/data-filters/) to "Filter out errors known to be caused by browser extensions."
Screen Shot 2022-04-28 at 3 27 40 PM

**Comment by Simek** [comment]
Refs: https://github.com/expo/styleguide/pull/19

After looking into this issue, I have spotted that the root cause of the error flooding could potentially be outside the docs code. To confirm this, we need to merge in the PR mentioned above, release new `styleguide` version, update the package in the docs dependencies and deploy the website.

**Comment by danstepanov** [comment]
@Simek what was the result of your merged PR?

**Comment by Simek** [comment]
@danstepanov It has been deployed yesterday, because we have a problem with `styleguide` earlier. Unfortunately, I'm not able to confirm if this helps, without the Sentry access.
